src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,309 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use Countable;
+use IteratorAggregate;
+use JsonSerializable;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONCodecLibrary;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_filter;
+use function array_key_exists;
+use function array_keys;
+use function array_map;
+use function array_values;
+use function count;
+use function is_array;
+use function is_numeric;
+use function iterator_to_array;
+use function max;
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON array.
+ *
+ * The internal data will be filtered through array_values() during BSON
+ * serialization to ensure that it becomes a BSON array.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<int, TValue>
+ * @template-implements IteratorAggregate<int, TValue>
+ */
+final class LazyBSONArray implements ArrayAccess, Countable, IteratorAggregate, JsonSerializable
+{
+    /** @var PackedArray<TValue> */
+    private PackedArray $bson;
+
+    /** @var array<int, TValue> */
+    private array $read = [];
+
+    /** @var array<int, bool> */
+    private array $exists = [];
+
+    /** @var array<int, TValue> */
+    private array $set = [];
+
+    /** @var array<int, true> */
+    private array $unset = [];
+
+    private bool $entirePackedArrayRead = false;
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy array.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /** @param PackedArray<TValue>|list<TValue>|null $input */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = PackedArray::fromPHP([]);
+        } elseif ($input instanceof PackedArray) {
+            $this->bson = $input;
+        } elseif (is_array($input)) {
```

Comment:
That's a good point. The question is where we want to draw the line beyond which we expect an array to be a list. `PackedArray` certainly requires a list, while the `LazyBSONArray` and `BSONArray` classes accept arrays with non-consecutive integer indexes. Your comment also highlights another issue: users creating a LazyBSONArray with such an array might not expect keys to be reindexed. Consider:
```php
$array = ['foo', 'bar'];
unset($array['foo']); // Creates a "hole"
var_dump($array[1]); // string(3) "bar"

$bsonArray = new LazyBSONArray($array);
var_dump($bsonArray[1]); // Undefined offset
```

However, it gets worse if users start unsetting keys then re-assigning them: https://3v4l.org/YWLoP. Considering all of that, it may be wisest to require the argument to be a list and accepting the fact that PHP arrays are indeed too weird to be called true arrays. WDYT?

Reply:
> users creating a LazyBSONArray with such an array might not expect keys to be reindexed. Consider:
> ```
> $array = ['foo', 'bar'];
> unset($array['foo']); // Creates a "hole"
> var_dump($array[1]); // string(3) "bar"
> ```

I think you meant to write `unset($array[0])` in the above example.

I was thinking about hows this compares with BSONArray, which allows gaps/holes throughout its lifetime and up to the point we call `array_values()` in `bsonSerialize()` or `jsonSerialize()`.

If we consider the case where a codec constructs LazyBSONArray, I suppose it would most likely (or always?) come from an BSON array, which is a true list. In that case, the `array_values()` on construction isn't necessary.

So I think the question to focus on is whether we want LazyBSONArray to differ from BSONArray when users construct it themselves using a non-list array. I think the similar class names alone is reason enough to try and preserve the same behavior, so I'd defer to using `array_values()` during serialization time and accept that this may _not_ be a list during its lifetime.

And I think it already is prone to going from a list to a non-list since we implement `offsetUnset()`.

Reply:
Updated the class to accept lists with gaps, but reject any arrays that contain non-numeric keys. The latter check is conditional based on `array_is_list` so we don't need to check keys for well-formed lists.

Reply:
You are right. I've added tests here and in `BSONArray` to assert against the order of elements when assigning them in reverse order. `LazyBSONArray` now invokes `sort` on newly set keys before iterating to ensure the order is correct. I didn't change `BSONArray`, but we can discuss how to proceed there. Take the following code:
```php
$array = new BSONArray(); // Alternatively, LazyBSONArray

$array[2] = 'foo';
$array[1] = 'bar';
$array[0] = 'baz';

var_dump(iterator_to_array($array));
```

As a user, I'd expect both to return `['baz', 'bar', 'foo']`, but `BSONArray` returns `['foo', 'bar', 'baz']`. `LazyBSONArray` behaves as expected with the latest changes.